pkg/controller/core/workload_controller.go

```diff
@@ -481,3 +493,74 @@ func (r *WorkloadReconciler) handleLimitsToRequests(wl *kueue.Workload) {
 		}
 	}
 }
+
+func (r *WorkloadReconciler) adjustResources(log logr.Logger, wl *kueue.Workload) {
+	r.handlePodOverhead(log, wl)
+	r.handlePodLimitRange(log, wl)
+	r.handleLimitsToRequests(wl)
+}
+
+type resourceUpdatesHandler struct {
+	r *WorkloadReconciler
+}
+
+func (h *resourceUpdatesHandler) Create(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+	//TODO: the eventHandler should get a context soon, and this could be dropped
+	// https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/handler/eventhandler.go
+	ctx := context.TODO()
+	log := ctrl.LoggerFrom(ctx)
+	log.V(5).Info("update resources event (create)")
+	h.handle(ctx, e.Object, q)
+}
+
+func (h *resourceUpdatesHandler) Update(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	ctx := context.TODO()
+	log := ctrl.LoggerFrom(ctx)
+	log.V(5).Info("update resources event (update)")
+	h.handle(ctx, e.ObjectNew, q)
+}
+
+func (h *resourceUpdatesHandler) Delete(e event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	ctx := context.TODO()
+	log := ctrl.LoggerFrom(ctx)
+	log.V(5).Info("update resources event (delete)")
+	h.handle(ctx, e.Object, q)
+}
+
+func (h *resourceUpdatesHandler) Generic(_ event.GenericEvent, _ workqueue.RateLimitingInterface) {
+}
+
+func (h *resourceUpdatesHandler) handle(ctx context.Context, obj client.Object, q workqueue.RateLimitingInterface) {
+	switch v := obj.(type) {
+	case *corev1.LimitRange:
+		log := ctrl.LoggerFrom(ctx).WithValues("limitRange", klog.KObj(v))
+		ctx = ctrl.LoggerInto(ctx, log)
+		h.queueReconcileForPending(ctx, q, client.InNamespace(v.Namespace))
+	case *nodev1.RuntimeClass:
+		log := ctrl.LoggerFrom(ctx).WithValues("runtimeClass", klog.KObj(v))
+		ctx = ctrl.LoggerInto(ctx, log)
+		h.queueReconcileForPending(ctx, q, client.MatchingFields{indexer.WorkloadRuntimeClassKey: v.Name})
+	default:
+		panic(v)
+	}
+}
+
+func (h *resourceUpdatesHandler) queueReconcileForPending(ctx context.Context, q workqueue.RateLimitingInterface, opts ...client.ListOption) {
+	log := ctrl.LoggerFrom(ctx)
+	lst := kueue.WorkloadList{}
+	opts = append(opts, client.MatchingFields{indexer.WorkloadAdmittedKey: string(metav1.ConditionFalse)})
+	err := h.r.client.List(ctx, &lst, opts...)
```

Comment:
Here is one case where it could be problematic: in between the time you list a workload here and the end of this function, a workload could be deleted. With this implementation, the workload would remain in the queue at the end.

Maybe we need a method: `UpdateWorkloadIfExists` in the queue. Although what could end up happening is that the workload remains in the queue and it's eventually considered for admission. I would expect the API call for the admission to fail, because the Workload doesn't exist. So maybe it's fine.

But I'll leave it up to you to consider these kinds of scenarios.

Reply:
Can you add an integration test that:
1. creates a workload
2. updates limit range
3. deletes workload

No waiting between the steps. Have two scenarios: one where the workload fits, one where it doesn't.

We need to check that the ClusterQueue doesn't have any admitted or pending workloads by the end of the test.

> I think this is acceptable, The same can happen if the workload is deleted during the scheduling (nominate and friends)

I think you are right, but an integration tests would add confidence and protect against future bugs.

Reply:
done